My solution of the GCDandLCM exercise

Maksim-Yeager opened the pull request:
If there is a problem, it might be the code documentation.

Maksim-Yeager committed:
My solution of the GCDandLCM exercise

If there is a problem, it might be the code documetation

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,33 @@
-public class Main {
-    public static void main(String[] args) {
+public class GCD {
```
The class is missing a javadoc.

ivorusev replied:
Still missing a javaoc here. Classes should have one as well, the same was as methods

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,33 @@
+public class GCD {
 
+    /**
+     * Method to find GCD
```
The comment violates the java conventions. Method names start with a verb and end with a dot. I don't need to know from the comment that it is a method so this is redundant. For instance good comment would be `Calculates the greatest common divisor of two numbers.`.

ivorusev replied:
You could add to the comment that you are using Euclidean algorithm for instance.

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,33 @@
+            }
+            return a;
+        } else {
+            throw new IllegalArgumentException("Invalid numbers");
```
This is good, but I would swap the if the other way around. This way the code will be less nested in brackets.

ivorusev replied:
Why is the number invalid? Shouldn't the exception message provide some more information?

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,33 @@
+    /**
+     * Method to find GCD
+     */
+    public static int GCD(int a, int b) {
```
a and b are bad for parameter names. Better would be to have something like firstNumber, secondNumber

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,33 @@
     }
+
+    /**
+     * Method to find LCM
```
Same issue with the javadoc as the one above.

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,33 @@
+    /**
+     * Method to find LCM
+     */
+    public static int LCM(int a, int b) {
```
Method name violates the naming conventions. Do you know why?

Maksim-Yeager replied:
Method names should start with a lowercase letter

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,33 @@
+        if (a != 0 || b != 0) {
+            return (a * b) / GCD(a, b);
+        } else {
+            throw new IllegalArgumentException("Invalid numbers");
```
Again I'd swap the if.

ivorusev commented on introducing-java/src/test/java/com/onto/javacourse/intro/gcd/GCDTest.java:
```diff
@@ -0,0 +1,20 @@
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class GCDTest {
```
This is good but you have an if statement in both methods. Can you also test the exception throwing, see https://www.baeldung.com/junit-assert-exception

Maksim-Yeager committed:
Refactoring of GCD and LCM methods, improved comments following Java conventions, improved exception handling, added JUnit tests, and organized code for better readability and adherence to best practices.

GCD Class Methods:

Updated comments for the GCD and LCM methods to adhere to Java conventions.
Reversed the order of conditions in the LCM method to reduce nested brackets.

Comments:

Improved comments for better readability and adherence to conventions.

Exception Handling:

Refactored exception handling in both GCD and LCM methods for more consistency and clarity.
Tests:

Added JUnit tests for the GCD and LCM methods.

Code Structure:

Organized the code and ensured it follows Java conventions.

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,42 @@
+    public static int calculateGcd(int firstNumber, int secondNumber) {
+        if (firstNumber == 0 || secondNumber == 0) {
+            throw new IllegalArgumentException("Invalid numbers, they cannot be 0.");
+        } else {
```
My idea was to remove the else clause, you can do that when throwing an exception

```
        if (firstNumber == 0 || secondNumber == 0) {
            throw new IllegalArgumentException("Invalid numbers, they cannot be 0.");
        }
        while ....
```

ivorusev commented on introducing-java/src/main/java/com/onto/javacourse/intro/gcd/GCD.java:
```diff
@@ -1,5 +1,42 @@
+    public static int calculateLcm(int firstNumber, int secondNumber) {
+        if (firstNumber == 0 || secondNumber == 0) {
+            throw new IllegalArgumentException("Invalid numbers, they cannot be 0.");
+        } else {
```
Same as above, the else can be removed.

ivorusev commented on introducing-java/src/test/java/com/onto/javacourse/intro/gcd/GCDTest.java:
```diff
@@ -0,0 +1,66 @@
+        assertTrue(actualMessageBothZero.contains(expectedMessage));
+        assertTrue(actualMessageFirstIsZero.contains(expectedMessage));
+        assertTrue(actualMessageSecondIsZero.contains(expectedMessage));
+
```
You don't need this new line. New lines are typically used to split logical parts in the code and this one is redundant.

Maksim-Yeager committed:
Updated the GCD class with comprehensive Javadoc comments

I also removed the else in the methods
